zcash_client_backend/src/data_api.rs

```diff
@@ -262,3 +351,233 @@ impl ShieldedOutput for DecryptedOutput {
         None
     }
 }
+
+#[cfg(feature = "test-dependencies")]
+pub mod testing {
+    use std::collections::HashMap;
+
+    use zcash_primitives::{
+        block::BlockHash,
+        consensus::BlockHeight,
+        merkle_tree::{CommitmentTree, IncrementalWitness},
+        note_encryption::Memo,
+        primitives::{Nullifier, PaymentAddress},
+        sapling::Node,
+        transaction::{components::Amount, Transaction, TxId},
+        zip32::ExtendedFullViewingKey,
+    };
+
+    use crate::{
+        address::RecipientAddress,
+        decrypt::DecryptedOutput,
+        proto::compact_formats::CompactBlock,
+        wallet::{AccountId, SpendableNote, WalletTx},
+    };
+
+    use super::{error::Error, BlockSource, ShieldedOutput, WalletRead, WalletWrite};
+
+    pub struct MockBlockSource {}
+
+    impl BlockSource for MockBlockSource {
+        type Error = Error<u32>;
+
+        fn with_blocks<F>(
+            &self,
+            _from_height: BlockHeight,
+            _limit: Option<u32>,
+            _with_row: F,
+        ) -> Result<(), Self::Error>
+        where
+            F: FnMut(CompactBlock) -> Result<(), Self::Error>,
+        {
+            Ok(())
+        }
+    }
+
+    pub struct MockWalletDB {}
```

Comment:
Non-blocking: This should probably be implemented as an in-memory wallet, in order to ensure that the various `WalletRead` APIs return data consistent with any wallet writes.

Reply:
Yup! That's what I'm going to do in my open pairing today.